### ISSUE TITLE
HTTP request routing appears broken for EE10

### ISSUE THREAD

**Issue by MuffinTheMan** [GoogleCloudPlatform/appengine-java-standard#316]
Our organization is attempting to migrate our Appengine Standard Java 21 application from EE8 to EE10; and we're currently blocked.

I hope there's something simple I'm doing incorrectly, but I've spent more hours than I'd care to admit trying to figure this out (reading documentation, asking ChatGPT, experimenting, etc). I cannot get routing to work as I expect per Jakarta EE10 documentation. To make this completely replicatable, I created a bare-bones Java 21 EE10 Appengine web application that anyone can checkout and deploy to an Appengine Standard project. I'll lay out some of of the issues in this issue, but the README in this repository lists more details with actual examples of some of the behavior I'm experiencing: https://github.com/MuffinTheMan/appengine-standard-ee10-sample

To summarize, none of the `<error-page>` mappings in `web.xml` appear to work when deployed (though they work locally). Additionally, using an `HttpServlet` to forward a request to a JSP or static file doesn't work when deployed (though this works locally as well). For example, this code in an `HttpServlet::doGet` (or `::doPost`) method never resolves (the browser never receives a response and just hangs eternally):
```
req.getRequestDispatcher("/WEB-INF/test.jsp").forward(req, resp);
```

1) There appears to be bugs in the deployed Appengine runtime
2) The local Appengine server behaves differently from the deployed version (which makes troubleshooting a nightmare)

**Comment by ludoch** [comment]
@lachlan-roberts 

Thanks for the repro case, we will study it.

**Comment by ludoch** [comment]
I see JDO being involved. I am not sure this has been maintained up to EE10 and Java 21...

Is it currently working for 21 and EE8?

**Comment by MuffinTheMan** [comment]
> I see JDO being involved. I am not sure this has been maintained up to EE10 and Java 21...
> 
> Is it currently working for 21 and EE8?

If I'm being honest, I don't know what the JDO config is for specifically in our case 😅--I left it there since we had it before (I just changed the namespace from `javax` to `jakarta`). We've noticed no issues so far with 21 and EE8, which has been on our pre-production environment for a few weeks now.

**Comment by MuffinTheMan** [comment]
Thank you, @lachlan-roberts! This may be separate, so let me know if I should open another issue; but I also can't get a catch-all `Filter` to work (which we use extensively in our actual application). This doesn't work either locally or when deployed. I pushed another commit with an example `TestFilter` that I would expect should log every request, but it only logs once on application startup: https://github.com/MuffinTheMan/appengine-standard-ee10-sample/commit/f38c476d3a0b915dfc880727e0f8c7ff198e3457

Note: I also tried `@WebFilter("/jsp")`, which should map to requests handled by [JspServlet](https://github.com/MuffinTheMan/appengine-standard-ee10-sample/blob/master/src/com/test/JspServlet.java), but it's never executed (even when hitting `/jsp`, which properly renders `test.jsp`.

**Comment by lachlan-roberts** [comment]
@MuffinTheMan I cannot reproduce any issue with this, your `TestFilter` seems to be completely working for me.

Can you try testing with the latest changes from the main branch.
https://github.com/GoogleCloudPlatform/appengine-java-standard/blob/main/TRYLATESTBITSINPROD.md

**Comment by MuffinTheMan** [comment]
@lachlan-roberts, I can't seem to figure out how to adapt those steps to my Gradle setup (I'm not really familiar with Maven and am by no means a Gradle expert 😬). I can't seem to figure out where to place the jars so they're in the right spot when deployed to Appengine. I get these two successive logs in GCP:
```
INFO [pid1-app] Starting app (pid 11): /bin/sh -c exec java --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio.charset=ALL-UNNAMED --add-opens java.base/java.util.concurrent=ALL-UNNAMED --add-opens java.logging/java.util.logging=ALL-UNNAMED -showversion -XX:+PrintCommandLineFlags -Djava.class.path=runtime-main.jar -Dclasspath.runtimebase=.: com/google/apphosting/runtime/JavaRuntimeMainWithDefaults --fixed_application_path=. . [session:CBK9GQ3]
```

```
WARNING [pid1-app] Exit 1 app (pid 11): /bin/sh -c exec java --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio.charset=ALL-UNNAMED --add-opens java.base/java.util.concurrent=ALL-UNNAMED --add-opens java.logging/java.util.logging=ALL-UNNAMED -showversion -XX:+PrintCommandLineFlags -Djava.class.path=runtime-main.jar -Dclasspath.runtimebase=.: com/google/apphosting/runtime/JavaRuntimeMainWithDefaults --fixed_application_path=. . [session:CBK9GQ3]
```

Here's a screenshot showing where the jars are in my build:
<img width="515" alt="Screenshot 2024-12-05 at 11 28 36 AM" src="https://github.com/user-attachments/assets/52de1642-9fb0-489a-a350-50b7544a1c98">

**Comment by ludoch** [comment]
Hi, in the final output, these 6 jars should be 1 level up, not in the lib/ area, but siblings of lib/ directory.

**Comment by MuffinTheMan** [comment]
@ludoch, here? This doesn't work either 😞
![Screenshot 2024-12-05 at 2 35 55 PM](https://github.com/user-attachments/assets/9c3d8b56-d5cd-4b9c-ace3-186cbc45f92f)

**Comment by ludoch** [comment]
My bad!, sorry to have wasted your time...
The correct dir is still 1 up, so sibling of WEB-INF not inside WEB-INF

The key here is the "." directory defined in the entrypoint as -Dclasspath.runtimebase=.
and the current workind directory context is stage-app in prod (with a different name).
Try again with that.

**Comment by MuffinTheMan** [comment]
Thank you, @ludoch--this worked! I'll need to do more testing of the fix, but I'm at least able to run the unreleased fix version. For posterity (I'll try to organize things and open a PR with more details in `TRYLATESTBITSINPROD.md` at some point):
![Screenshot 2024-12-05 at 3 36 46 PM](https://github.com/user-attachments/assets/58ea4370-5f0e-4e80-bf95-0ecd52072723)

Snippet in `appengine-web.xml`:
```
<entrypoint>
    java
    -showversion
    -XX:+PrintCommandLineFlags
    --add-opens java.base/java.lang=ALL-UNNAMED
    --add-opens java.base/java.nio.charset=ALL-UNNAMED
    --add-opens java.base/java.util.concurrent=ALL-UNNAMED
    --add-opens java.logging/java.util.logging=ALL-UNNAMED
    -Dclasspath.runtimebase=.:
    -Djava.class.path=./runtime-main.jar
    -Djava.library.path=.:
    com/google/apphosting/runtime/JavaRuntimeMainWithDefaults
    --fixed_application_path=/workspace
    .
</entrypoint>
```
